### DISCUSSION TITLE
Implement Relation and Rollup column for Database Tables

### DISCUSSION THREAD

**Discussion by domenico-ruggiano** [toeverything/AFFiNE#8881] [Ideas]
Hello team, 

thank you for creating AFFiNE, I'm loving this tool for organizing and collaborating in my organization. I’d like to propose adding **Relation and Rollup columns**, which would allow users to create connections between data in multiple tables and leverage those relationships for deeper insights.  

### Relation Columns 
Relation columns allow users to link entries between tables.  
- Example: Connect a “Customers” table with an “Orders” table to see which customers placed which orders.  
- **Two-way relations:** Automatically update both sides of the connection. For example, if a customer is added to an order in the “Orders” table, the order will appear in the “Customers” table.  
- **Self-referencing relations:** Allow a table to reference its own entries, such as creating parent-child relationships within a “Tasks” table to track dependencies.  

### Rollup Columns

Rollup columns aggregate data from related tables based on relations.  
- Example: In the “Customers” table, calculate the total amount spent by each customer based on data from the related “Orders” table.  
- **Aggregation options:** Include sum, average, count, min, max, earliest date, latest date, and more.  

#### Use Cases  
1. **Project Management:** Link a “Tasks” table to a “Projects” table, allowing teams to see task breakdowns and project progress.  
2. **CRM:** Connect a “Clients” table to an “Interactions” table to track client activity and engagement.  
3. **Inventory Management:** Link a “Products” table to a “Sales” table to track stock levels and revenue.  
4. **Task Hierarchies:** Use self-referencing relations in a “Tasks” table to create parent-child task relationships and track dependencies.  

#### Benefits to Users  
- **Streamlined Data Organization:** Relating tables reduces duplication and improves efficiency.  
- **Dynamic Data Insights:** Rollups provide real-time summaries and calculations for better decision-making.  
- **Expanded Use Cases:** Makes AFFiNE applicable to complex workflows, including project management, CRM, and inventory tracking.  

I’d love to hear your thoughts and collaborate on refining this idea further.

**Comment by zigarot** [comment]
Yes Please,
I use a table for games with it's own fields, but I want the relational to the developers, which has it's own fields, contacts etc. this is also then linked to a publisher/strategist. This allows me to find details from a publisher that might have multiple developers that have multiple games.

**Comment by Lorex-ia** [comment]
Piggybacking what has been said above. You guys are doing amazing work !

**Comment by justinpik** [comment]
This two-way sync (or transclusion) is a must have feature for me and for others who will use databases (e.g., project management, PKM & second brain uses) often - Notion and Loop do this in some degrees. Without this feature, there is double and triple work and data integrity issues that result.  If Table B (e.g. "Projects") is related to Table A (e.g. "Tasks") because some of the Tasks in Table A relate to various projects in Table B, the user should be able to make changes to A table from B table (update Task due date, start date, etc.) without having to flip back to table A. It's not enough to be able to display results from a related table - the user doesn't want to have to click into it to make changes then flip back.  Otherwise love the product and direction. I feel like I'm in limbo between wanting to use Affine now and wanting to buy-in 100%, but I and many others need this feature.

**Comment by gavinmcfall** [comment]
+1 — this is the single biggest gap for anyone migrating off Notion.

I just moved a large personal/work Notion setup onto self-hosted AFFiNE: ~40 databases and ~1,250 rows across three workspaces. Almost everything came across beautifully (the block model + edgeless canvas are fantastic) — **except the relations.** I had several relation-heavy clusters that I was forced to flatten to plain text or drop entirely:

- A 6-table "racing scene" graph: drivers ↔ clubs ↔ venues ↔ events ↔ rulesets, including **self-references** (peer links, club affiliations, ruleset supersedes).
- A People ↔ Assessments / Goals / Teams / Work-History "career hub" — one Teams table alone had **10 relation columns** pointing into People.
- A filament-catalogue ↔ print-profiles ↔ materials cluster.
- A task database with parent/child + "blocked by" **self-relations**.

Without relation columns, all that cross-linking is lost — you end up with disconnected tables and manual text references that drift out of sync. Rollups are the natural follow-on, but **relations — especially two-way and self-referencing — are the must-have** that would let AFFiNE fully replace Notion for structured data.

Big +1, and happy to help test whenever this lands on the roadmap.

**Comment by darkskygit** [comment]
We will advance and refine this feature in the next major version(0.28)
Converted to issue #15157 , so close this discussion
